server/src/main/java/org/elasticsearch/rest/action/search/HttpChannelTaskHandler.java

```diff
@@ -0,0 +1,158 @@
+                }
+
+                private void unlink(Task task) {
+                    //the synchronized blocks are to make sure that only link or unlink for a specific task can happen at a given time,
+                    //they can't happen concurrently. The link flag is needed because unlink can still be called before link, which would
+                    //lead to piling up task ids that are never removed from the map.
+                    //It may look like only either synchronized or the flag are needed but they both are. In fact, the flag is needed to
+                    //ensure that we don't link a task if we have already unlinked it. But it's not enough as, once we start the linking,
+                    //we do want its corresponding unlinking to happen, but only once the linking is completed. With only
+                    //the flag, we would just miss unlinking for some tasks that are being linked while onResponse is called.
+                    synchronized(task) {
+                        try {
+                            //nothing to do if link was not called yet: we would not find the task anyways.
+                            if (link.getAndIncrement() > 0) {
+                                CloseListener closeListener = httpChannels.get(httpChannel);
+                                TaskId taskId = new TaskId(client.getLocalNodeId(), task.getId());
+                                closeListener.unregisterTask(taskId);
+                            }
+                        } catch(Exception e) {
+                            listener.onFailure(e);
+                        }
+                    }
+                }
+            });
+
+        CloseListener closeListener = httpChannels.computeIfAbsent(httpChannel, channel -> new CloseListener(client));
+        synchronized (task) {
+            //the task will only be registered if it's not completed yet, meaning if its TaskListener has not been called yet.
+            //otherwise, given that its listener has already been called, the task id would never be removed.
+            if (link.getAndIncrement() == 0) {
+                TaskId taskId = new TaskId(client.getLocalNodeId(), task.getId());
+                closeListener.registerTask(httpChannel, taskId);
+            }
+        }
+
+        //TODO test case where listener is registered, but no tasks have been added yet:
+        // - connection gets closed, channel will be removed, no tasks will be cancelled
+
+        //TODO check that no tasks are left behind through assertions at node close
+    }
+
+    final class CloseListener implements ActionListener<Void> {
+        private final Client client;
+        final Set<TaskId> taskIds = new CopyOnWriteArraySet<>();
+        private final AtomicReference<HttpChannel> channel = new AtomicReference<>();
+
+        CloseListener(Client client) {
+            this.client = client;
+        }
+
+        void registerTask(HttpChannel httpChannel, TaskId taskId) {
```

Comment:
We synchronize on the task when calling this function but different tasks can reference the same channel concurrently (http pipelining) so we need to synchronize this function too ?

Reply:
even if we use AtomicReference for the channel and CopyOnWriteArraySet for the taskids?